File: firebase/functions/lib/events/live_meetings/create_live_stream.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../on_call_function.dart';
import '../../utils/infra/firestore_utils.dart';
import 'mux_client.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/community/membership.dart';

class CreateLiveStream extends OnCallMethod<CreateLiveStreamRequest> {
  CreateLiveStream()
      : super(
          'CreateLiveStream',
          (jsonMap) => CreateLiveStreamRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    CreateLiveStreamRequest request,
    CallableContext context,
  ) async {
    if (context.authUid == null) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }

    final membershipDoc =
        'memberships/${context.authUid}/community-membership/${request.communityId}';
    final communityMembershipDoc =
        await firestore.document(membershipDoc).get();

    if (!communityMembershipDoc.exists) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }

    final membership = Membership.fromJson(
      firestoreUtils.fromFirestoreJson(communityMembershipDoc.data.toMap()),
    );

    if (!membership.isMod) {
      print('member not moderator: $membershipDoc');
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }

    // Create mux live stream
    print('creating a livestream');
    final liveStream = await muxApi.createLiveStream();
    print(liveStream);

    return CreateLiveStreamResponse(
      muxId: liveStream['id'],
      muxPlaybackId: (liveStream['playback_ids'] as List<dynamic>)
          .firstWhere((entry) => entry['policy'] == 'public')['id'],
      streamServerUrl: 'rtmp://global-live.mux.com:5222/app',
      streamKey: liveStream['stream_key'],
    ).toJson();
  }
}

File: firebase/functions/src/events/live_meetings/create_live_stream.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { muxApi } from './mux_client';
import { Membership } from '../../types';

interface CreateLiveStreamRequest {
  communityId: string;
}

interface CreateLiveStreamResponse {
  muxId?: string;
  muxPlaybackId?: string;
  streamServerUrl?: string;
  streamKey?: string;
}

export class CreateLiveStream extends OnCallMethod<CreateLiveStreamRequest> {
  constructor() {
    super('CreateLiveStream', (jsonMap) => jsonMap as CreateLiveStreamRequest);
  }

  async action(
    request: CreateLiveStreamRequest,
    context: functions.https.CallableContext
  ): Promise<CreateLiveStreamResponse> {
    const membershipSnap = await firestore
      .doc(`memberships/${context.auth?.uid}/community-membership/${request.communityId}`)
      .get();
    const membership = firestoreUtils.fromFirestoreJson(membershipSnap.data() ?? {}) as unknown as Membership;

    if (!(membership as any).isAdmin) {
      console.log(`member not admin: memberships/${context.auth?.uid}/community-membership/${request.communityId}`);
      throw new functions.https.HttpsError('failed-precondition', 'unauthorized');
    }

    console.log('creating a livestream');
    const liveStream = await muxApi.createLiveStream();
    console.log(liveStream);

    const ls = liveStream as Record<string, unknown>;
    return {
      muxId: ls['id'] as string,
      muxPlaybackId: (ls['playback_ids'] as Array<Record<string, string>>).find(
        (entry) => entry['policy'] === 'public'
      )?.['id'],
      streamServerUrl: 'rtmp://global-live.mux.com:5222/app',
      streamKey: ls['stream_key'] as string,
    };
  }
}
